File: packages/wnode-sdk-ts/test/vrf.test.ts
import { WnodeClient } from '../src/client';
import { VRFClient } from '../src/vrf';

describe('VRFClient Simulation', () => {
  let client: WnodeClient;
  let vrf: VRFClient;

  beforeEach(() => {
    client = new WnodeClient({
      endpoint: 'http://localhost',
      chainId: 1,
      sdkVersion: '1.0.0',
      apiVersion: '1.0',
    });
    vrf = new VRFClient(client);
  });

  it('should simulate fulfillment and return proof', async () => {
    const result = await vrf.simulateFulfillment({
      request: {
        vrfCoordinator: '0x123',
        keyHash: '0xabc',
        subscriptionId: 1,
        requestConfirmations: 3,
        callbackGasLimit: 100000,
        numWords: 1,
      },
      blockTag: 'finalized',
    });

    expect(result.simulatedOutput).toBeDefined();
    expect(result.simulatedOutput.to).toBe('0x123');
    expect(result.proof).toBeDefined();
    expect(result.proof.version).toBe('1.0');
    expect(result.proof?.blockTag).toEqual({ finalized: true });
  });
});

File: wnode-sdk-go/vrf_test.go
package sdk_test

import (
	"testing"
	"github.com/wnodeltd/wnode/wnode-sdk-go"
)

func TestVRFClient_SimulateFulfillment(t *testing.T) {
	client := sdk.NewWnodeClient(sdk.WnodeClientConfig{})
	vrf := sdk.NewVRFClient(client)

	params := sdk.SimulateFulfillmentParams{
		Request: sdk.GenerateVRFRequestParams{
			VRFCoordinator:       "0x123",
			KeyHash:              "0xabc",
			SubscriptionID:       "1",
			RequestConfirmations: 3,
			CallbackGasLimit:     100000,
			NumWords:             1,
		},
		BlockTag: sdk.BlockTag{Finalized: true},
	}

	result, err := vrf.SimulateFulfillment(params)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.SimulatedOutput.To != "0x123" {
		t.Fatalf("Expected target 0x123, got %s", result.SimulatedOutput.To)
	}

	if result.Proof == nil {
		t.Fatalf("Expected proof to be populated")
	}

	if result.Proof.Version != "1.0" {
		t.Fatalf("Expected proof version 1.0, got %s", result.Proof.Version)
	}
}
